ports/nrf/modules/machine/uart.c

```diff
@@ -319,8 +320,17 @@ static mp_uint_t mp_machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t
 
 static mp_uint_t mp_machine_uart_write(mp_obj_t self_in, const void *buf_in, mp_uint_t size, int *errcode) {
     machine_uart_obj_t *self = self_in;
+    uint8_t *buf = (uint8_t *)buf_in;
+    #if !NRFX_UART_ENABLED
+    if (!nrfx_is_in_ram(buf_in)) {
+        // EasyDMA requires that transfer buffers are placed in DataRAM,
+        // NRFX_ERROR_INVALID_ADDR if the are not.
+        buf = alloca(size);
```

Comment:
This seems pretty dangerous, the stack may overflow if the buffer is large.

Maybe limit it to 256 bytes alloca, and do a loop over the buffer to break it up and send it in chunks?

Reply:
Good catch! I had been thinking this was really just stuff like exception messages where it's sending const strings from the firmware. But looking forward, certainly when RomFS comes in there will be plenty of opportunity to send data directly from flash.
I've reworked this in a chunked send loop which was tested on a nrf52840 dev board (repl on uart).